### PR TITLE
chore: Use `pass_through` instead of `strict` in `narwhals.from_native`, use stable.v1 API more

### PR DESCRIPTION
**Pull request by MarcoGorelli**
I'm addressing the API mix noted in https://github.com/vega/altair/pull/3690

Also, seeing as `1.13.1` is now the minimum, you can now use `pass_through` instead of `strict` (https://narwhals-dev.github.io/narwhals/backcompat/#after-stablev1)

### CODE REVIEW

File: pyproject.toml [hunk 1]
```diff
@@ -34,7 +34,7 @@ keywords = [
 ]
 requires-python = ">=3.9"
 dynamic = ["version"]
-license-files = { paths = ["LICENSE"] }
+license = {file = "LICENSE"}
```
Comment (MarcoGorelli) [inline pyproject.toml:37]: drive-by fix which the latest ruff release flags

File: tests/vegalite/v5/test_params.py [hunk 2]
```diff
@@ -107,7 +107,7 @@ def test_parameter_naming():
     # test automatic naming which has the form such as param_5
     prm0, prm1, prm2 = (alt.param() for _ in range(3))
 
-    res = re.match("param_([0-9]+)", prm0.param.name)
+    res = re.match(r"param_([0-9]+)", prm0.param.name)
```
Comment (MarcoGorelli) [inline tests/vegalite/v5/test_params.py:110]: also to address the latest ruff release

File: pyproject.toml (hunk 1 above)
Comment (dangotbanned) [reply pyproject.toml:37]: Related
- https://github.com/vega/altair/discussions/3687#discussioncomment-11351453

Comment (MarcoGorelli) [reply pyproject.toml:37]: thanks!

i took this from https://packaging.python.org/en/latest/guides/writing-pyproject-toml/#license

**Comment by dangotbanned** [comment]
@MarcoGorelli do you know if we could use a `ruff` rule for the `narwhals` imports?

If we had something like this, but suggesting the `stable.v1` path it would be quite handy

https://github.com/vega/altair/blob/936c0af6f473bea113fbf162c444a8a99a571447/pyproject.toml#L393-L400

File: altair/utils/core.py [hunk 3]
```diff
@@ -471,7 +470,7 @@ def sanitize_narwhals_dataframe(
     # See https://github.com/vega/altair/issues/1027 for why this is necessary.
     local_iso_fmt_string = "%Y-%m-%dT%H:%M:%S"
     for name, dtype in schema.items():
-        if dtype == nw.Date and nw.get_native_namespace(data) is get_polars():
+        if dtype == nw.Date and nw.dependencies.is_polars_dataframe(data):
```
Comment (dangotbanned) [inline altair/utils/core.py:473]: `pyright` isn't happy with the typeguard here:

```
Type "polars.dataframe.frame.DataFrame | narwhals.stable.v1.DataFrame[TIntoDataFrame@sanitize_narwhals_dataframe]" is not assignable to return type "DataFrame[TIntoDataFrame@sanitize_narwhals_dataframe]"
  Type "polars.dataframe.frame.DataFrame | narwhals.stable.v1.DataFrame[TIntoDataFrame@sanitize_narwhals_dataframe]" is not assignable to type "DataFrame[TIntoDataFrame@sanitize_narwhals_dataframe]"
    "DataFrame" is not assignable to "DataFrame[TIntoDataFrame@sanitize_narwhals_dataframe]"PylancereportReturnType
```

Reporting on https://github.com/MarcoGorelli/altair/blob/308cb7a1d8c29a38d05239609b1b76d95bf6f614/altair/utils/core.py#L493

```py
return data.select(columns)
```

Comment (MarcoGorelli) [reply altair/utils/core.py:473]: that's so cool! pyright helped prevent a bug, because I should've passed `data.to_native()` here

now, to figure out why the tests are still passing, maybe test coverage needs increasing - taking a look 👀

Comment (MarcoGorelli) [reply altair/utils/core.py]: Looks like this wasn't explicitly being tested for - I've fixed up, and have added a test

Wow

I'll keep this example handy for the next time someone tells me that Python typing is useless and doesn't help prevent bugs 😄

File: altair/utils/core.py [hunk 4]
```diff
@@ -471,7 +470,7 @@ def sanitize_narwhals_dataframe(
     # See https://github.com/vega/altair/issues/1027 for why this is necessary.
     local_iso_fmt_string = "%Y-%m-%dT%H:%M:%S"
     for name, dtype in schema.items():
-        if dtype == nw.Date and nw.get_native_namespace(data) is get_polars():
+        if dtype == nw.Date and nw.dependencies.is_polars_dataframe(data.to_native()):
```
Comment (dangotbanned) [inline altair/utils/core.py:473]:
```suggestion
    local_iso_fmt_string = "%Y-%m-%dT%H:%M:%S"
    is_polars = nw.dependencies.is_polars_dataframe(data.to_native())
    for name, dtype in schema.items():
        if dtype == nw.Date and is_polars:
```

File: pyproject.toml [hunk 5]
```diff
@@ -398,6 +398,15 @@ Use `Union[T, None]` instead.
...
 """
+"narwhals.dependencies".msg = """
+Import `dependencies` from `narwhals.stable.v1` instead.
+"""
+"narwhals.typing".msg = """
+Import `typing` from `narwhals.stable.v1` instead.
+"""
+"narwhals.dtypes".msg = """
+Import `dtypes` from `narwhals.stable.v1` instead.
+"""
```
Comment (dangotbanned) [inline pyproject.toml:409]: Perfect, thanks @MarcoGorelli 

Can't wait for this to tell me off in https://github.com/vega/altair/pull/3631/files

File: altair/utils/core.py (hunk 3 above)
Comment (MarcoGorelli) [reply altair/utils/core.py]: and thank you so much for spotting this

this sends chills down my spine 🥶

**Review by dangotbanned — APPROVED** [review]
Thanks again @MarcoGorelli

File: pyproject.toml (hunk 5 above)
Comment (dangotbanned) [reply pyproject.toml:409]: @MarcoGorelli I'm really happy with this, honestly might be something you wanna suggest to do in the docs

```cmd
>>> ruff check

altair\datasets\_cache.py:8:1: TID251 `narwhals.dependencies` is banned: Import `dependencies` from `narwhals.stable.v1` instead.

  |
7 | import narwhals.stable.v1 as nw
8 | from narwhals.dependencies import get_pyarrow
  | ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^ TID251
9 | from narwhals.typing import IntoDataFrameT, IntoFrameT
  |

altair\datasets\_cache.py:9:1: TID251 `narwhals.typing` is banned: Import `typing` from `narwhals.stable.v1` instead.

   |
 7 | import narwhals.stable.v1 as nw
 8 | from narwhals.dependencies import get_pyarrow
 9 | from narwhals.typing import IntoDataFrameT, IntoFrameT
   | ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^ TID251
10 |
11 | from altair.datasets._typing import VERSION_LATEST
   |

altair\datasets\_loader.py:5:1: TID251 `narwhals.typing` is banned: Import `typing` from `narwhals.stable.v1` instead.

  |
3 | from typing import TYPE_CHECKING, Generic, final, overload
4 |
5 | from narwhals.typing import IntoDataFrameT, IntoFrameT
  | ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^ TID251
6 |
7 | from altair.datasets._readers import _Reader, backend
  |

altair\datasets\_readers.py:33:1: TID251 `narwhals.typing` is banned: Import `typing` from `narwhals.stable.v1` instead.

   |
32 | import narwhals.stable.v1 as nw
33 | from narwhals.typing import IntoDataFrameT, IntoExpr, IntoFrameT
   | ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^ TID251
34 |
35 | from altair.datasets._cache import DatasetCache
   |

tests\test_datasets.py:15:1: TID251 `narwhals.dependencies` is banned: Import `dependencies` from `narwhals.stable.v1` instead.

   |
14 |   import pytest
15 | / from narwhals.dependencies import (
16 | |     is_into_dataframe,
17 | |     is_pandas_dataframe,
18 | |     is_polars_dataframe,
19 | |     is_pyarrow_table,
20 | | )
   | |_^ TID251
21 |   from narwhals.stable import v1 as nw
   |

Found 5 errors.

```
